homeassistant/components/arduino.py

```diff
@@ -0,0 +1,183 @@
+from homeassistant.helpers.entity import ToggleEntity
+from homeassistant.const import (STATE_ON, STATE_OFF, DEVICE_DEFAULT_NAME)
+
+
+DOMAIN = "arduino"
+DEPENDENCIES = []
+BOARD = None
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup(hass, config):
+    """ Setup the Arduino component. """
+
+    # pylint: disable=global-statement
+    if not validate_config(config,
+                           {DOMAIN: ['port']},
+                           _LOGGER):
+        return False
+
+    try:
+        from PyMata.pymata import PyMata
+
+    except ImportError:
+        _LOGGER.error("Error while importing dependency pyMata.")
+
+        return False
+
+    global BOARD
+    BOARD = ArduinoBoard(config[DOMAIN]['port'])
+
+    return True
+
+
+class ArduinoBoard():
+    """ Represents an Arduino board within Home Assistant. """
+
+    def __init__(self, port):
+        try:
+            from PyMata.pymata import PyMata
+
+        except ImportError:
+            _LOGGER.error("Error while importing dependency pyMata.")
+
+        self._port = port
+        self._board = PyMata(self._port, verbose=False)
+
+    def set_mode(self, pin, direction, mode):
+        """ Sets the mode and the direction of a given pin. """
+        if mode == 'analog' and direction == 'in':
+            self._board.set_pin_mode(pin,
+                                     self._board.INPUT,
+                                     self._board.ANALOG)
+            self._board.enable_analog_reporting(pin)
+        elif mode == 'analog' and direction == 'out':
+            self._board.set_pin_mode(pin,
+                                     self._board.OUTPUT,
+                                     self._board.ANALOG)
+        elif mode == 'digital' and direction == 'in':
+            self._board.set_pin_mode(pin,
+                                     self._board.OUTPUT,
+                                     self._board.DIGITAL)
+            self._board.enable_digital_reporting(pin)
+        elif mode == 'digital' and direction == 'out':
+            self._board.set_pin_mode(pin,
+                                     self._board.OUTPUT,
+                                     self._board.DIGITAL)
+        elif mode == 'pwm':
+            self._board.set_pin_mode(pin,
+                                     self._board.OUTPUT,
+                                     self._board.PWM)
+
+    def digital_out_high(self, pin):
+        """ Sets a given digital pin to high. """
+        self._board.digital_write(pin, 1)
+
+    def digital_out_low(self, pin):
+        """ Sets a given digital pin to low. """
+        self._board.digital_write(pin, 0)
+
+    def get_firmata(self):
+        """ Return the version of the Firmata firmware. """
+        return self._board.get_firmata_version()
+
+    def disconnect(self):
+        """ Disconnects the board and closes the serial connection. """
+        self._board.close
+
+
+class ArduinoDeviceABC(ToggleEntity):
```

Comment:
I see that this class keeps track if a pin is on/off by keeping track of it itself instead of asking the arduino what the actual value is. The documentation does seem to have a [`digital_read`](https://htmlpreview.github.io/?https://github.com/MrYsLab/PyMata/blob/master/documentation/html/PyMata.pymata.PyMata-class.html#digital_read) method. Could that be used?

Reading it from the arduino would make sure that when a write fails to change the pin we do not report the wrong state.